HTTP/3 proxy CONNECT and MASQUE CONNECT-UDP support for OpenSSL QUIC

aritrbas opened the pull request:
This adds (1) HTTP/3 proxy CONNECT support and (2) MASQUE CONNECT-UDP support for OpenSSL QUIC stack.

### Usage:
- `curl -k -4 --http1.1 --proxy-insecure --proxytunnel --proxy-http3 --proxy https://<proxy> https://<target>`
- `curl -k -4 --http2 --proxy-insecure --proxytunnel --proxy-http3 --proxy https://<proxy> https://<target>`
- `curl -k -4 --http3-only --proxy-insecure --proxyudptunnel --proxy https://<proxy> https://<target>`
- `curl -k -4 --http3-only --proxy-insecure --proxyudptunnel --proxy-http2 --proxy https://<proxy> https://<target>`
- `curl -k -4 --http3-only --proxy-insecure --proxyudptunnel --proxy-http3 --proxy https://<proxy> https://<target>`

### References:
[RFC 9297- HTTP Datagrams and the Capsule Protocol](https://www.rfc-editor.org/rfc/rfc9297)
[RFC 9298 - Proxying UDP in HTTP](https://www.rfc-editor.org/rfc/rfc9298)

bagder commented on lib/capsule.c:
```diff
@@ -0,0 +1,310 @@
+#include "curl_memory.h"
+#include "memdebug.h"
+
+uint64_t curl_capsule_ntohll(uint64_t value)
```
We use `curl_` prefix for public functions and `Curl_` for global private functions, so I presume this should be `Curl_`...

bagder commented on docs/libcurl/symbols-in-versions:
```diff
@@ -688,6 +688,7 @@ CURLOPT_HTTPGET                 7.8.1
 CURLOPT_HTTPHEADER              7.1
 CURLOPT_HTTPPOST                7.1           7.56.0
 CURLOPT_HTTPPROXYTUNNEL         7.3
+CURLOPT_HTTPPROXYUDPTUNNEL      8.16.0
```
Minor detail but the first release this can go into is currently 8.17.0

bagder commented on include/curl/curl.h:
```diff
@@ -796,16 +796,18 @@ typedef CURLcode (*curl_ssl_ctx_callback)(CURL *curl,    /* easy handle */
-#define CURLPROXY_SOCKS5_HOSTNAME 7L /* Use the SOCKS5 protocol but pass along
+#define CURLPROXY_SOCKS5          6L /* added in 7.10 */
+#define CURLPROXY_SOCKS4A         7L /* added in 7.18.0 */
+#define CURLPROXY_SOCKS5_HOSTNAME 8L /* Use the SOCKS5 protocol but pass along
```
You must not the change the values of existing defines here (as that breaks the ABI). You can *add* new ones though.

aritrbas replied:
Ack, fixed in latest commit.

bagder commented on lib/curlx/dynbuf.c:
```diff
@@ -229,7 +229,7 @@ CURLcode curlx_dyn_addf(struct dynbuf *s, const char *fmt, ...)
   va_list ap;
   DEBUGASSERT(s);
-  DEBUGASSERT(s->init == DYNINIT);
+  /* DEBUGASSERT(s->init == DYNINIT); */
```
Why? This looks like you're covering for a mistake?

aritrbas replied:
This was done for debugging an issue and I forgot to revert it, fixed in the latest commit.

bagder commented on docs/libcurl/curl_easy_setopt.md:
```diff
@@ -471,6 +471,10 @@ See CURLOPT_HTTPPOST(3)
 
+## CURLOPT_HTTPPROXYUDPTUNNEL
+
+UDP Tunnel through the HTTP proxy. CURLOPT_HTTPPROXYUDPTUNNEL(3)
```
Can we elaborate on this? When or why would a user want to enable this and tunnel UDP explicitly?

bagder commented:
Wouldn't it make sense to have at least these tests as a start?

- HTTP/3 successfully proxied over a HTTP/3 proxy
- HTTP/3 unsuccessfully proxied over a HTTP/3 proxy (proxy doesn't support)
- HTTP/3 successfully proxied over a HTTP/2 proxy
- HTTP/3 unsuccessfully proxied over a HTTP/2 proxy (proxy doesn't support)
- HTTP/3 successfully proxied over a HTTP/1 proxy
- HTTP/3 unsuccessfully proxied over a HTTP/1 proxy (proxy doesn't support)
- HTTP/2 successfully proxied over a HTTP/3 proxy
- HTTP/2 unsuccessfully proxied over a HTTP/3 proxy (proxy doesn't support)
- HTTP/1 successfully proxied over a HTTP/3 proxy
- HTTP/1 unsuccessfully proxied over a HTTP/3 proxy (proxy doesn't support)

bagder commented:
Now this PR has merge conflicts. Any news on the questions?

aritrbas commented on lib/capsule.c:
```diff
@@ -0,0 +1,310 @@
+
+#include "curl_setup.h"
+
+#if !defined(CURL_DISABLE_PROXY) && !defined(CURL_DISABLE_HTTP)
```
Yes, fixed in the latest commit.

aritrbas commented:
> What's your thoughts and plans for tests for these new features?

All the testing I had done prior to raising the PR involved manual testing on Ubuntu22.04 using the following setups:
- client (cURL) <-----tunnel-----> proxy (h2o) <------> server (h2o)
- client (cURL) <-----tunnel-----> proxy (h2o) <------> server (nginx)
- client (cURL) <-----tunnel-----> proxy (envoy) <------> server (h2o)
- client (cURL) <-----tunnel-----> proxy (envoy) <------> server (nginx)
 
> The OpenSSL-QUIC backend is still marked experimental in curl while the ngtcp2 one is not. How come you focus your implementation on that?

I was more familiar with the OpenSSL code (such as the use of OpenSSL BIO) and never worked on the ngtcp2 stack before 😅. That's why I picked this up on the OpenSSL-QUIC backend. I can try making the changes on ngtcp2, but given my unfamiliarity with the code, it might take some time to get it in a working state.

> Wouldn't it make sense to have at least these tests as a start?

Yes, I have added these tests for a start.
I was facing some issues while trying to use the existing `nghttpx/httpd` infra for adding these tests. I looked at the[ nghttpx documentation](https://nghttp2.org/documentation/nghttpx-howto.html#http-3) and it looks like `nghttpx does not support HTTP/3 on backend connection.`  So, I have added the tests using [h2o ](https://h2o.examp1e.net/configure/http3_directives.html) in [this commit](https://github.com/curl/curl/pull/18331/commits/e4ec6222e8701a1a499974254a27adeb60dc5874). It still leverages most of the existing http pytest infra, just that it involves an extra dependency on h2o now (not sure if that's the right way to go about it)

> Now this PR has merge conflicts. Any news on the questions?

Addressed the merge conflicts and updated the basic tests.

> We greatly appreciate the initial efforts shown here, but we cannot accept this code like this. It feels like a wasted effort, but this is too much too advanced code to merge without assurances, tests and adjustments, If there is no feedback on this soon, this PR will be closed without action. Sadly.

Sorry for the delayed response on this, got busy with multiple other things. As things stand now, I still need to make some fixes to limit the code to OpenSSL specific HTTP/3 builds - that's the cause of majority of the CI failures now on this branch. Following that, I will try to look at porting the changes to the ngtcp2 backend and working on further review comments / testing requirements.

bagder commented:
> I was more familiar with the OpenSSL code (such as the use of OpenSSL BIO) and never worked on the ngtcp2 stack before 😅. That's why I picked this up on the OpenSSL-QUIC backend. I can try making the changes on ngtcp2, but given my unfamiliarity with the code, it might take some time to get it in a working state.

I think it would be useful to focus the efforts there. I think there is a non-zero risk that OpenSSL-QUIC ends up a dead end and never becomes production ready in curl. Also for performance reasons.

> I have added the tests using h2o

That seems like a reasonable path to go down, at least for the moment. It should support the necessary mechanisms. We just need to work a little to get that baby into the CI builds as well. (It could be worth noticing however that h2o performs worse than some of the other HTTP/3 servers in raw transfer performance but I don't think it should matter for this testing)

bagder commented on lib/capsule.c:
```diff
@@ -0,0 +1,312 @@
+    capsule_length = capsule_decode_varint(&decode_ptr);
+
+    if(capsule_length == HTTP_INVALID_VARINT) {
+      infof(data, "Error! Invalid varint length encoding");
```
```suggestion
      failf(data, "Error! Invalid varint length encoding");
```

bagder commented on lib/capsule.c:
```diff
@@ -0,0 +1,312 @@
+      break;
+
+    if(*context_id) {
+      infof(data, "Error! Invalid context ID: %02x", *context_id);
```
```suggestion
      failf(data, "Error! Invalid context ID: %02x", *context_id);
```

bagder commented on lib/capsule.c:
```diff
@@ -0,0 +1,312 @@
+                            bytes_to_read, &bytes_read);
+
+      if(bytes_read != bytes_to_read) {
+        infof(data, "Error! Read less than expected %zu %zu",
```
```suggestion
        failf(data, "Error! Read less than expected %zu %zu",
```

bagder commented on lib/capsule.c:
```diff
@@ -0,0 +1,312 @@
+
+    /* Verify we read the entire capsule */
+    if(remaining_capsule_length > 0) {
+      infof(data, "Error! Could not fit entire capsule: remaining=%zu",
```
```suggestion
      failf(data, "Error! Could not fit entire capsule: remaining=%zu",
```

bagder commented on lib/cf-h1-proxy.c:
```diff
@@ -165,7 +174,11 @@ static void h1_tunnel_go_state(struct Curl_cfilter *cf,
   case H1_TUNNEL_ESTABLISHED:
     CURL_TRC_CF(data, cf, "new tunnel state 'established'");
-    infof(data, "CONNECT phase completed");
+    if(cf->conn->bits.udp_tunnel_proxy)
+      infof(data, "CONNECT-UDP phase completed for HTTP proxy");
+    else
+      infof(data, "CONNECT phase completed for HTTP proxy");
+
```
```suggestion
    infof(data, "CONNECT%s phase completed for HTTP proxy".
          cf->conn->bits.udp_tunnel_proxy ? "-UDP": "");
```

bagder commented on lib/cf-h3-proxy.c:
```diff
@@ -0,0 +1,3038 @@
+        nread = nghttp3_conn_read_stream(ctx->h3.conn, s->id,
+                                         buf, blen, 0);
+        if(nread < 0) {
+          failf(data, "nghttp3_conn_read_stream(len=%zu) error: %s",
```
As it sets result to CURLE_OK below here, should this perhaps be a plain infof() instead?

bagder commented on lib/cf-h3-proxy.c:
```diff
@@ -0,0 +1,3038 @@
+      ctx->items_max = 0;
+      tmpptr = realloc(ctx->poll_items, nmax * sizeof(SSL_POLL_ITEM));
+      if(!tmpptr) {
+        free(ctx->poll_items);
+        ctx->poll_items = NULL;
```
```suggestion
        Curl_safefree(ctx->poll_items);
```

bagder commented on lib/cf-h3-proxy.c:
```diff
@@ -0,0 +1,3038 @@
+
+      tmpptr = realloc(ctx->curl_items, nmax * sizeof(struct Curl_easy *));
+      if(!tmpptr) {
+        free(ctx->curl_items);
+        ctx->curl_items = NULL;
```
```suggestion
        Curl_safefree(ctx->curl_items);
```

bagder reviewed (commented):
Some suggestions!

vszakats commented on lib/cf-h1-proxy.c:
```diff
@@ -220,11 +234,20 @@ static CURLcode start_CONNECT(struct Curl_cfilter *cf,
 
-  infof(data, "Establish HTTP proxy tunnel to %s", req->authority);
+  if(cf->conn->bits.udp_tunnel_proxy)
+    infof(data, "Establishing HTTP proxy UDP tunnel to %s:%s",
+                      data->state.up.hostname, data->state.up.port);
```
```suggestion
    infof(data, "Establishing HTTP proxy UDP tunnel to %s:%s",
          data->state.up.hostname, data->state.up.port);
```

vszakats commented on lib/cf-h1-proxy.c:
```diff
@@ -278,6 +301,50 @@ static CURLcode send_CONNECT(struct Curl_cfilter *cf,
+  struct SingleRequest *k = &data->req;
+
+  if(checkprefix("Transfer-Encoding:", header)) {
+    if(Curl_compareheader(header,
+                           STRCONST("Transfer-Encoding:"),
+                           STRCONST("chunked"))) {
+      CURL_TRC_CF(data, cf, "CONNECT-UDP Response --> "
+                  "Transfer-Encoding: chunked");
+      ts->chunked_encoding = TRUE;
+      /* reset our chunky engine */
+      Curl_httpchunk_reset(data, &ts->ch, TRUE);
+    }
+  }
+  else if(checkprefix("Capsule-protocol:", header)) {
+    if(Curl_compareheader(header,
+                           STRCONST("Capsule-protocol:"),
+                           STRCONST("?1"))) {
+      CURL_TRC_CF(data, cf, "CONNECT-UDP Response --> Capsule-protocol: ?1");
+    }
+  }
+  else if(Curl_compareheader(header,
+                              STRCONST("Connection:"), STRCONST("close"))) {
+    ts->close_connection = TRUE;
+    CURL_TRC_CF(data, cf, "CONNECT-UDP Response --> Connection: close");
+  }
+  else if(!strncmp(header, "HTTP/1.", 7) &&
+           ((header[7] == '0') || (header[7] == '1')) &&
+           (header[8] == ' ') &&
+           ISDIGIT(header[9]) && ISDIGIT(header[10]) && ISDIGIT(header[11]) &&
+           !ISDIGIT(header[12])) {
+    /* store the HTTP code from the proxy */
+    data->info.httpproxycode = k->httpcode = (header[9] - '0') * 100 +
+                          (header[10] - '0') * 10 + (header[11] - '0');
```
```suggestion
    if(Curl_compareheader(header,
                          STRCONST("Transfer-Encoding:"),
                          STRCONST("chunked"))) {
      CURL_TRC_CF(data, cf, "CONNECT-UDP Response --> "
                  "Transfer-Encoding: chunked");
      ts->chunked_encoding = TRUE;
      /* reset our chunky engine */
      Curl_httpchunk_reset(data, &ts->ch, TRUE);
    }
  }
  else if(checkprefix("Capsule-protocol:", header)) {
    if(Curl_compareheader(header,
                          STRCONST("Capsule-protocol:"),
                          STRCONST("?1"))) {
      CURL_TRC_CF(data, cf, "CONNECT-UDP Response --> Capsule-protocol: ?1");
    }
  }
  else if(Curl_compareheader(header,
                             STRCONST("Connection:"), STRCONST("close"))) {
    ts->close_connection = TRUE;
    CURL_TRC_CF(data, cf, "CONNECT-UDP Response --> Connection: close");
  }
  else if(!strncmp(header, "HTTP/1.", 7) &&
          ((header[7] == '0') || (header[7] == '1')) &&
          (header[8] == ' ') &&
          ISDIGIT(header[9]) && ISDIGIT(header[10]) && ISDIGIT(header[11]) &&
          !ISDIGIT(header[12])) {
    /* store the HTTP code from the proxy */
    data->info.httpproxycode = k->httpcode = (header[9] - '0') * 100 +
                               (header[10] - '0') * 10 + (header[11] - '0');
```
indent

vszakats commented on lib/cf-h3-proxy.c:
```diff
@@ -0,0 +1,3038 @@
+  case SSL_ERROR_WANT_ASYNC:
+    return "SSL_ERROR_WANT_ASYNC";
+#endif
+#ifdef SSL_ERROR_WANT_ASYNC_JOB
```
```suggestion
#ifdef SSL_ERROR_WANT_ASYNC_JOB  /* OpenSSL 1.1.0+, LibreSSL 3.6.0+ */
```

vszakats commented on lib/cf-h3-proxy.c:
```diff
@@ -0,0 +1,3038 @@
+  return Curl_capsule_process_udp(cf, data, &proxy_ctx->inbufq, buf, len, err);
+}
+
+static CURLcode cf_h3_proxy_recv(struct Curl_cfilter *cf,
+                                struct Curl_easy *data,
+                                char *buf, size_t len, size_t *pnread)
```
```suggestion
static CURLcode cf_h3_proxy_recv(struct Curl_cfilter *cf,
                                 struct Curl_easy *data,
                                 char *buf, size_t len, size_t *pnread)
```

vszakats commented on lib/cf-h3-proxy.c:
```diff
@@ -0,0 +1,3038 @@
+}
+
+static CURLcode cf_h3_proxy_query(struct Curl_cfilter *cf,
+                                   struct Curl_easy *data,
+                                   int query, int *pres1, void *pres2)
```
```suggestion
                                  struct Curl_easy *data,
                                  int query, int *pres1, void *pres2)
```
